### DISCUSSION TITLE
read response from BLE

### DISCUSSION THREAD

**Discussion by syntaxxxxx** [JuulLabs/kable#557] [Q&A] [answered]
hi all

i am writing to write payload request with json with response

payload request is sent dan device received the data, but the problem is, i got response ??? like this when i read response

this is my implementation code

    private suspend fun sendRequest(body: String = ""): String {
        withContext(Dispatchers.IO) {
            peripheral.write(request, body.toByteArray(Charsets.UTF_8), WriteType.WithResponse)
        }

        peripheral.observe(request).firstOrNull()

        return peripheral.read(response).toString(Charsets.UTF_8).also {
            Timber.d("BLE woy -> Response $it")
        }
    }
    
    thanks!

**Comment by twyatt** [comment, marked as answer]
The following discussion may be applicable/helpful? https://github.com/JuulLabs/kable/discussions/507

**Reply by syntaxxxxx** [reply]
thanks in advance! @twyatt

**Reply by syntaxxxxx** [reply]
I tried using this, but I encountered a problem,

Characteristic 6e400005-b5a3-f393-e0a9-e50e24daaa9e with writeWithoutResponse property not found

if using the previous approach is works but the response is truncated

**Reply by twyatt** [reply]
What do you mean the response is truncated?
If you provide some logs and sample code that would be helpful.

**Reply by syntaxxxxx** [reply]
private suspend fun sendRequest(body: String = ""): ByteArray {
        withContext(Dispatchers.IO) {
            peripheral.write(request, body.toByteArray(Charsets.UTF_8), WriteType.WithResponse)
        }

        return peripheral.read(response).also {
            Timber.d("BLE woy -> Response $it")
        }
        
        // Output for ByteArray -> [@bada

**Reply by syntaxxxxx** [reply]
```kotlin
private val UART_SERVICE_UUID = UUID.fromString("6E411111-B5A5-F393-E0A9-E50E24DCCA9E")
private val TX_CHARACTERISTIC_UUID = UUID.fromString("6E411112-B5A5-F393-E0A9-E50E24DCCA9E")
private val RX_CHARACTERISTIC_UUID = UUID.fromString("6E411113-B5A5-F393-E0A9-E50E24DCCA9E")

private val request = characteristicOf(
    UART_SERVICE_UUID.toString(),
    TX_CHARACTERISTIC_UUID.toString()
)

private val response = descriptorOf(
    UART_SERVICE_UUID.toString(),
    RX_CHARACTERISTIC_UUID.toString(),
    "00002902-0000-1000-8000-00805f9b34fb"
)
```

**Reply by twyatt** [reply]
> // Output for ByteArray -> [@bada

Are you referring to the `[@bada` as indication that it is truncated? The `toString` for `ByteArray` doesn't represent its contents, if that's what you were assuming? You'll want to convert it using your desired encoding, or you could inspect the "raw" data using [`toHexString`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-hex-string.html).

Additionally, your example code could potentially miss packets, you should execute your `read` before your `write`.

**Reply by syntaxxxxx** [reply]
i got the point

thanks a lot! @twyatt

**Comment by syntaxxxxx** [comment]
hi travis! @twyatt 

i back again

 i still got response ??? like this when i read response

this is my implementation code

     peripheral.read(response).also {
         println("execute lah 5 ${byteArrayToString(it)}")
     }

    private suspend fun writeExampleValue(peripheral: Peripheral) {
        val value = 1
        val data = byteArrayOf(value.toByte())
        peripheral.write(request, data = data, WriteType.WithResponse)
    }

**Reply by syntaxxxxx** [reply]
<img width="516" alt="Screenshot 2023-11-16 at 13 02 36" src="https://github.com/JuulLabs/kable/assets/33598887/20ebf8ed-d3ec-4f51-8cfc-92ee97f19998">

**Reply by twyatt** [reply]
It really depends on the character encoding of the data you're reading. It might be helpful to see the hex representation of what data you're getting back, to see if the data seems reasonable?

```kotlin
peripheral.read(response).also {
    println("execute lah 5 ${it.toHexString()}")
}
```

Alternatively, Kable provides logging for you that you can enable (and can be configured to show hex representations of all I/O):

```kotlin
val peripheral = scope.peripheral(advertisement) {
    logging {
        level = Data
    }
}
```

See https://github.com/JuulLabs/kable#logging for more details.

**Reply by syntaxxxxx** [reply]
i got data: 00 00 from onDescriptorRead from logging

**Reply by twyatt** [reply]
Are you intending to read the descriptor value (vs. a characteristic)?
What data are you expecting?
Hex of `00 00` means you read 2 bytes of value `0`.

**Reply by syntaxxxxx** [reply]
the goal is

when i send payload request json to device via BLE i want to response back in json form to

mean while i tried to using android api BLE, it worked, when i use kable library, it didn't work, I think there is something wrong with my implementation when using the kable, could you help me?

**Reply by twyatt** [reply]
Do you have the source publicly available for when you used the Android API directly (and it worked for you)? That would help me understand how you previously interacted with your peripheral and help me to provide you with a corresponding Kable implementation.

**Reply by syntaxxxxx** [reply]
```kotlin
    private val callback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(
            gatt: BluetoothGatt,
            status: Int,
            newState: Int,
        ) {
            super.onConnectionStateChange(gatt, status, newState)

            btGatt = gatt

            when (newState) {
                BluetoothGatt.STATE_CONNECTING -> {
                    Timber.d("UART onConnectionStateChange STATE_CONNECTING")
                    _connectionState.value = ConnectionState.CONNECTING
                }

                BluetoothGatt.STATE_CONNECTED -> {
                    Timber.d("UART onConnectionStateChange STATE_CONNECTED")
                    _connectionState.value = ConnectionState.CONNECTED
                    gatt.requestMtu(512)
                }

                BluetoothGatt.STATE_DISCONNECTING -> {
                    Timber.d("UART onConnectionStateChange STATE_DISCONNECTING")
                    _connectionState.value = ConnectionState.DISCONNECTING
                }

                BluetoothGatt.STATE_DISCONNECTED -> {
                    Timber.d("UART onConnectionStateChange STATE_DISCONNECTED")
                    channelConnectionState.trySend(OnConnectionState(status, newState))

                    _connectionState.value = ConnectionState.DISCONNECTED
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt?, status: Int) {
            super.onServicesDiscovered(gatt, status)
            Timber.d("UART onServicesDiscovered $gatt and $status")

            btGatt?.services?.flatMap { it.characteristics }?.find { svcChar ->
                Timber.d("UART onServicesDiscovered -> ${svcChar.uuid.toString().uppercase()}")
                svcChar.uuid.toString().uppercase() == "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"
            }?.also {
                if (status == BluetoothGatt.GATT_SUCCESS) {
                    val service: BluetoothGattService? = gatt?.getService(UART_SERVICE_UUID)

                    txCharacteristic = service?.getCharacteristic(TX_CHARACTERISTIC_UUID) ?: return
                    rxCharacteristic = service.getCharacteristic(RX_CHARACTERISTIC_UUID)

                    gatt.setCharacteristicNotification(rxCharacteristic, true)

                    val descriptor = rxCharacteristic.getDescriptor(
                        UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
                    )

                    descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    gatt.writeDescriptor(descriptor)
                }
            }
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic,
        ) {
            super.onCharacteristicChanged(gatt, characteristic)
            Timber.d("UART onCharacteristicChanged $gatt $characteristic")

            val newMaxLength = 512
            val responseData = ByteArray(newMaxLength)

            if (characteristic == rxCharacteristic) {
                val value = characteristic.value
                value.copyInto(responseData)

                val response = String(responseData, Charsets.UTF_8)
                Timber.d("UART onCharacteristicChanged response $response")
                _response.value = response
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt?, mtu: Int, status: Int) {
            super.onMtuChanged(gatt, mtu, status)
            Timber.d("UART onMtuChanged $gatt, $mtu, $status")
            btGatt?.discoverServices()
        }
    }
```

@twyatt Here my implementation code using android native API

**Reply by twyatt** [reply]
Does your peripheral send you data on connect, or does it send data in response to you sending it something?
I don't see any place in your code where you write data (i.e. `txCharacteristic` is never used)?

**Reply by syntaxxxxx** [reply]
```kotlin
    override fun sendData(command: String): Flow<Boolean> = flow {
        Timber.d("UART sendData $command")

        if (::txCharacteristic.isInitialized) {
            txCharacteristic.value = command.toByteArray()
            val status = btGatt?.writeCharacteristic(txCharacteristic)
        } else {
            Timber.d("UART sendData write TXchar uartCharacteristic has not been initialized")
            emit(false)
        }
    }
```

@twyatt this, how i write data and use txCharacteristic

**Reply by twyatt** [reply]
Can you share where you call `sendData`? I'm trying to understanding timings of when you read vs. write data.

**Reply by syntaxxxxx** [reply]
viewModelScope.launch {
            bleGatt.sendData(request).collect { result ->
            }
        }

@twyatt here, i call sendData

**Reply by twyatt** [reply]
I _think_ your issue was that you [were defining your **receive** characteristic](https://github.com/JuulLabs/kable/discussions/557#discussioncomment-6858308) using `descriptorOf` (rather than `characteristicOf`).

There are many ways to tackle this, but I think the following would be a simple approach:

```kotlin
fun CoroutineScope.example(advertisement: Advertisement) =
    ExamplePeripheral(this, advertisement)

class ExamplePeripheral internal constructor(
    scope: CoroutineScope,
    advertisement: Advertisement,
) {

    private val peripheral = scope.peripheral(advertisement) {
        onServicesDiscovered {
            requestMtu(512)
        }
        logging {
            level = Data
        }
    }

    private val uart = "6E411111-B5A5-F393-E0A9-E50E24DCCA9E"
    private val tx = characteristicOf(
        service = uart,
        characteristic = "6E411112-B5A5-F393-E0A9-E50E24DCCA9E",
    )
    private val rx = characteristicOf(
        service = uart,
        characteristic = "6E411113-B5A5-F393-E0A9-E50E24DCCA9E",
    )

    private val incoming = peripheral
        .observe(rx)
        .shareIn(scope, started = Eagerly)

    suspend fun connect() = peripheral.connect()

    suspend fun send(command: String): String =
        incoming
            .onSubscription { peripheral.write(tx, command.encodeToByteArray()) }
            .first()
            .decodeToString()
}
```

Then, you can use it with:

```kotlin
val advertisement = scanner.advertisements.first()
val example = scope.example(advertisement).apply { connect() }
val response = example.send("example command")
println("response: $response")

// Some time later..
scope.cancel() // Cancel the scope when you're done w/ the peripheral (and want to dispose of it).
```

**Reply by syntaxxxxx** [reply]
@twyatt wah thanks for the answer, it worked

but my response is: execute lah 4 {"c_id":1,"c_str":"s

i didn't get full response, i suppose need requestMtu(512) right? if right, how i implement that if i make:

```kotlin
class ExamplePeripheral internal constructor(
    scope: CoroutineScope,
    advertisement: Advertisement
)
```

to shared module and consume in android and ios specific platform

**Reply by twyatt** [reply]
Ya, that is likely an MTU issue. I've updated my previous comment to include setting the MTU, but the relevant bit is:

```kotlin
private val peripheral = scope.peripheral(advertisement) {
    onServicesDiscovered {
        requestMtu(512)
    }
    logging {
        level = Data
    }
}
```

**Reply by twyatt** [reply]
> to shared module and consume in android and ios specific platform

I believe the code example I shared will work in `commonMain` source set.

[SensorTag sample](https://github.com/JuulLabs/sensortag) is a good example/reference for how to share Kable code w/ the supported platforms.

**Reply by syntaxxxxx** [reply]
yes it works in `commonMain` source set. but for `requestMtu(512)` Unresolved reference, how to solved that to support android and ios platform?

```
private val peripheral = scope.peripheral(advertisement) {
    onServicesDiscovered {
        requestMtu(512)
    }
    logging {
        level = Data
    }
}
```

Does code above can be shared?

**Reply by twyatt** [reply]
The MTU cannot be changed on iOS — I can't remember the exact value, but it is less than `512`. Something to be aware of.

As for how to have the code work w/ multiple platforms, one possible approach could be:

```kotlin
// common
internal expect fun PeripheralBuilder.configure()
private val peripheral = scope.peripheral(advertisement) {
    configure()
    logging {
        level = Data
    }
}

// Android
internal actual fun PeripheralBuilder.configure() {
    onServicesDiscovered {
        requestMtu(512)
    }
}

// iOS
internal actual fun PeripheralBuilder.configure() {
    // No-op
}
```

**Reply by syntaxxxxx** [reply]
hai travis @twyatt 

it works thanks in advance!

I have question. What possibilities can lead to Kable throwing a GattRequestRejectedException, other than:

* - Device is busy (i.e. a previous request is still in-progress)
 * - An Android internal failure occurred (i.e. an underlying [RemoteException] was thrown)

* @throws GattRequestRejectedException if underlying `BluetoothGatt` method call returns `false`.
* @throws GattStatusException if response has a non-`GATT_SUCCESS` status.

**Reply by twyatt** [reply]
> Device is busy (i.e. a previous request is still in-progress)

Kable has safe guards (per `Peripheral` object) to prevent executing multiple requests at once, so this _shouldn't_ happen. If you create multiple `Peripheral`s that represent the same physical peripheral (e.g. call `scope.peripheral(advertisement)` multiple times with the same `advertisement`) and then use them simultaneously then that would be skirting around the safe guards.

> What possibilities can lead to Kable throwing a GattRequestRejectedException, other than

As you referenced, there are a number of causes of that failure, but specifically looking from the perspective of the code, it is thrown at:

https://github.com/JuulLabs/kable/blob/f5bf08b4cbabbd3ea5c3c8601fb365d22b4c5718/core/src/androidMain/kotlin/Connection.kt#L70

Android's [`BluetoothGatt`](https://developer.android.com/reference/android/bluetooth/BluetoothGatt) describes for many of its functions that they return a boolean of:

> true, if the read operation was initiated successfully

Looking at the [`BluetoothGatt` source code](https://android.googlesource.com/platform/frameworks/base/+/9908112fd085d8b0d91e0562d32eebd1884f09a5/core/java/android/bluetooth/BluetoothGatt.java#919), it returns `false` when `RemoteException` is caught. I hadn't really dug deeper as to when Android internally throws `RemoteException`, I believe I just found various blog posts about it and noted what they identified as to when Android might return `false`, then then added those notes to Kable documentation.
